utils/s2n_random.c

```diff
@@ -390,7 +390,20 @@ int s2n_get_rdrand_data(struct s2n_blob *out)
             :"cc");
 #endif /* defined(__i386__) */
 
-            if (success) {
+            /* Some AMD CPUs will find that RDRAND "sticks" on all 1s but still reports success.
+             * Some other very old CPUs use all 0s as an error condition while still reporting success.
+             * If we encounter either of these suspicious values (a 1/2^63 chance) we'll treat them as
+             * a failure and generate a new value.
+             *
+             * In the future we could add CPUID checks to detect processors with these known bugs,
+             * however it does not appear worth it. The entropy loss is negligible and the
+             * corresponding likelihood that a healthy CPU generates either of these values is also
+             * negligible (1/2^63). Finally, adding processor specific logic would greatly
+             * increase the complexity and would cause us to "miss" any unknown processors with
+             * similar bugs. */
+            if (success &&
+                output.u64 != UINT64_MAX &&
+                output.u64 != 0) {
                 break;
             }
```

Comment:
The reports of the AMD bug are that it will generate all-ones over and over. Will this really help? We'll just get all ones again, eventually terminate the for loop when we run out of tries, but success will be "1". I think we need to set success back to 0 here in this branch. That way we'll trigger the S2N_ERR_RDRAND_FAILED error, which seems like our only option here.  If it's stuck on returning all-ones, there's nothing we can do to fix that. The machine needs to be rebooted. 

Separately for the all-zeroes case, if that is used as an error times and we want to ignore it, I don't think it's this simple. The i386 code might have returned all-zeroes for just one of the halves. So do we need to look at it more granularly? 

Also even though the bias we'd be introducing is tiny, as long as we have tries left, we could always combine bits from two "good" results to get a fair and unbiased result that could still represent all-zeroes. E.g. we could take the even bits from one roll, and the odd bits from another.

Reply:
Ah, missed setting the success flag back to 0.

Yes, we are screwed and need a reboot to resolve it gets stuck.  We will enter a mode of always failing to generate random data and will error out at the protocol later which is a preferable state than using deterministic random data for everything.

On the TLS  side, I can recall if we have a check on receiving all 1s in the secure random data.

I’ll add a check there too

Reply:
It looks to me like success is being set to `0`. Am I just missing where it isn't cleared?

In the AMD bug case, it will loop to max and just fail out. In the rare case that it really is a random all `0`s or all `1`s, it won't happen again (probabalistically), so you'll correctly succeed on the next iteration.

As for bias: If these values were being used directly then I might see the argument of removing the bias by XORing two values, however they are being used as additional entropy for the NIST DRBG which will mix them up so much anyway that I can't imagine it actually mattering.

Reply:
That’s my understanding too.